### ISSUE TITLE
Changing Focal Loss and Custom loss functions 

### ISSUE THREAD

**Issue by Miguel-P** [fizyr/keras-retinanet#710]
I'd like to use only the cross entropy loss, rather than focal loss in order to compare the difference in performance on my dataset. Everything trains and evaluates perfectly well when I use the train script as is. However, when I change the loss function that the model compiles with in bin/train.py, everything seems to break. 

I'm implementing these changes by altering the code, and reinstalling each time. Even the most basic change, namely:

- Not changing anything in models/\_\_init\_\_.py
- Not changing anything in the train script here:    

     `training_model.compile( loss={
              'regression'    : losses.smooth_l1()
              'classification': losses.focal(alpha=0.25, gamma=2.0)
     },
     optimizer=keras.optimizers.adam(lr=1e-5, clipnorm=0.001
    ))
    return model, training_model, prediction_model`

- only changing the way the focal loss is calculated in losses.py:
        `cls_loss = focal_weight * keras.backend.binary_crossentropy(labels, classification)`
to this:
        `cls_loss = keras.backend.binary_crossentropy(labels, classification)`
         `return keras.backend.sum(cls_loss) #/ normalizer`

Appears to break the model and results in absolutely no bounding boxes being predicted, despite the training process reporting a reducing (albeit large) classification loss.

Could anyone give some insight into why this is the case? Am I missing something deeper here with the way in which the model is compiled using the backbone argument or am I not understanding the way the classification loss is actually being applied to the model. Should the classification loss be normalized by the number of positive anchors regardless of the fact that I'm not using the focal loss?

Also, why does setting alpha and gamma = 1.0 also seem to result in breaking the model? The classification loss is still non zero and reducing as it should, but again, no bounding boxes are predicted at test time.

**Comment by yhenon** [comment]
>Appears to break the model and results in absolutely no bounding boxes being predicted, despite the training process reporting a reducing (albeit large) classification loss.

There is a huge ration of background anchors to actual objects, so the model learns to always predict bg. This is why the focal loss and normalization is present.

>why does setting alpha and gamma = 1.0 also seem to result in breaking the model

Alpha=1 will result in zero weighting for class -1, so the model won't work. Use `gamma=0` and `alpha=0.5` to simulate CE loss (expect poor results).
